uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/leave-request-popup-controller.js

```diff
@@ -0,0 +1,516 @@
+      $ctrl.calendar = {};
+      $ctrl.leaveRequestDayTypes = [];
+      $ctrl.balance = {
+        opening: 0,
+        change: {
+          amount: 0,
+          breakdown: []
+        },
+        closing: 0
+      };
+      $ctrl.period = {};
+      $ctrl.pagination = {
+        totalItems: $ctrl.balance.change.breakdown.length,
+        filteredbreakdown: $ctrl.balance.change.breakdown,
+        currentPage: 1,
+        numPerPage: 5,
+        pageChanged: function () {
+          //filter breakdowns
+          var begin = ((this.currentPage - 1) * this.numPerPage),
+            end = begin + this.numPerPage;
+
+          this.filteredbreakdown = $ctrl.balance.change.breakdown.slice(begin, end);
+        }
+      };
+      $ctrl.uiOptions = {
+        showDatePickerFrom: false,
+        showDatePickerTo: false,
+        isChangeExpanded: false,
+        datePickerOptions: {
+          startingDay: 1,
+          showWeeks: false
+        },
+        isAdmin: false,
+        userDateFormat: HR_settings.DATE_FORMAT,
+        showBalance: false,
+        multipleDays: true
+      };
+      $ctrl.error = undefined;
+      $ctrl.loading = {};
+
+      /**
+       * Change handler for change request type like multiple or single. It will
+       * reset dates, day types, change balance.
+       */
+      $ctrl.changeInNoOfDays = function () {
+        reset();
+        //reinitialize opening balance
+        $ctrl.balance.opening = $ctrl.selectedAbsenceType.remainder;
+      };
+
+      /**
+       * Whenever the absence type changes, update the balance opening.
+       * Also the balance change needs to be recalculated, if the `from` and `to`
+       * dates have been already selected
+       */
+      $ctrl.onAbsenceTypeChange = function () {
+        $ctrl.leaveRequest.type_id = $ctrl.selectedAbsenceType.id;
+        // get the `balance` of the newly selected absence type
+        $ctrl.balance.opening = $ctrl.selectedAbsenceType.remainder;
+
+        if ($ctrl.leaveRequest.from_date && $ctrl.leaveRequest.to_date) {
+          $ctrl.loading.calculateBalanceChange = true;
+          $ctrl.calculateBalanceChange().then(function () {
+            $ctrl.loading.calculateBalanceChange = false;
+          });
+        }
+      };
+
+      /**
+       * This should be called whenever a date has been changed
+       *
+       * First it syncs `from` and `to` date, if it's in 'single day' mode
+       * Then, if all the dates are there, it gets the balance change
+       *
+       */
+      $ctrl.onDateChange = function (date, isFrom) {
+        if ($ctrl.uiOptions.multipleDays) {
+          if ($ctrl.uiOptions.toDate && $ctrl.uiOptions.fromDate) {
+            $ctrl.uiOptions.showBalance = true;
+          }
+        } else {
+          if ($ctrl.uiOptions.fromDate) {
+            $ctrl.uiOptions.showBalance = true;
+          }
+        }
+
+        $ctrl.filterLeaveRequestDayTypes(date, !!isFrom)
+          .then(function () {
```

Comment:
calculateBalanceChange needs both day types to be defined for multiple days, it can change based on the date selected and therefore is in the callback, not parallel

similarly the setting of to date/types depends on whether its request for single/multiple days and which day type is selected